### PR TITLE
Qrcode auth

### PR DESCRIPTION
**Pull request by EionRobb**
Add support for authenticating with a QR Code instead of doing funky stuff with tokens

### CODE REVIEW

**Comment by EionRobb** [comment]
Probably should update the docs too to say that leaving the password blank will do the QR Code auth.  I wonder if the default `make` should build with QR auth support or not....

File: discord_rsa.c [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+#if !defined USE_MBEDTLS_CRYPTO && !defined USE_OPENSSL_CRYPTO && !defined USE_NSS_CRYPTO && !defined USE_GCRYPT_CRYPTO
+// #	ifdef _WIN32
+// #		define USE_WIN32_CRYPTO
+// #	else
+#		define USE_NSS_CRYPTO
+// #	endif
+#endif
+
+
+// Some of this is pinched from the gowhatsapp plugin
```
Comment (grimmy) [inline discord_rsa.c:34]: there's no whatsapp here?

File: Makefile [hunk 2]
```diff
@@ -30,7 +30,7 @@ CFLAGS ?= -O2 -g -pipe -Wall
 CFLAGS += -std=c99 -DDISCORD_PLUGIN_VERSION='"$(PLUGIN_VERSION)"' -DMARKDOWN_PIDGIN
 
 # Comment out to disable localisation
-CFLAGS += -DENABLE_NLS
+CFLAGS += -DENABLE_NLS
```
Comment (grimmy) [inline Makefile:33]: trailing space

File: discord_rsa.c [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+static void
+discord_null_cb() {
+}
+
+static void
+discord_display_qrcode(PurpleConnection *pc, const gchar *qr_code_raw, const gchar *qrcode_utf8, const guchar *image_data, gsize image_data_len)
+{
+    DiscordAccount *da = purple_connection_get_protocol_data(pc);
+	PurpleRequestUiOps *ui_ops = purple_request_get_ui_ops();
```
Comment (grimmy) [inline discord_rsa.c:47]: nit: mixed whitespace, there's a bunch in this function not got to comment on all of it.

File: discord_rsa.c [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+	g_free(secondary);
+	
+}
+
+static gchar *
+discord_base64_make_urlsafe(gchar *inout)
+{
+	//basically - and _ replace + and /
+	purple_util_chrreplace(inout, '+', '-');
+	purple_util_chrreplace(inout, '/', '_');
```
Comment (grimmy) [inline discord_rsa.c:111]: nice hack!

File: discord_rsa.c [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+		// No offsets
+		0, 0, 0, 0,
+		// Dimensions
+		realwidth & 0xFF, (realwidth/256) & 0xFF, realwidth & 0xFF, (realwidth/256) & 0xFF,
+		// 32 bits per pixel
+		32,
+		// "Origin in upper left-hand corner"
+		32
+	};
+	g_string_append_len(out, (const gchar *)tga_header, sizeof(tga_header));
```
Comment (grimmy) [inline discord_rsa.c:233]: for future reference, gdk-pixbuf just depends on glib, libjpeg, libpng, and libtiff.

File: discord_rsa.c [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+	SECKEYPublicKey *pubKey = g_dataset_get_data(da, "pubkey");
+	if (!pubKey) {
+		return NULL;
+	}
+
+	SECItem *cert_der = PK11_DEREncodePublicKey(pubKey);
+	
+	// gchar *b64crt = NSSBase64_EncodeItem(NULL, NULL, 0, cert_der);
+	// purple_str_strip_char(b64crt, '\n');
+	// purple_str_strip_char(b64crt, '\r');
```
Comment (grimmy) [inline discord_rsa.c:366]: guessing these can be deleted now.

File: libdiscord.c [hunk 7]
```diff
@@ -5038,6 +5069,136 @@ discord_close(PurpleConnection *pc)
...
+			
+			discord_display_qrcode(da->pc, qrcode_url, qrcode_utf8, qrcode_image, qrcode_image_len);
+			
+			g_free(qrcode_url);
+			g_free(qrcode_image);
+			g_free(qrcode_utf8);
+			
+		} else if (purple_strequal(op, "pending_finish")) {
+			// the app scanned, and is just confirming everything is OK
+
```
Comment (grimmy) [inline libdiscord.c:5166]: you could verify the user id and avatar hash and all that here, but not strictly necessary.

**Review by grimmy — APPROVED** [review]
Looks great!! Comments are all minor/nits.

File: discord_rsa.c (hunk 4 above)
Comment (EionRobb) [reply discord_rsa.c:110]: :D

File: discord_rsa.c (hunk 5 above)
Comment (EionRobb) [reply discord_rsa.c:233]: Trying to have as few dependencies as possible (specifically for console-based UIs such as bitlbee).  The TGA format is simple enough that it skips all the other deps, and still works with gdk-pixbuf

File: libdiscord.c (hunk 7 above)
Comment (EionRobb) [reply libdiscord.c:5166]: could do, but I'm lazy  :)

File: discord_rsa.c (hunk 5 above)
Comment (grimmy) [reply discord_rsa.c:233]: makes sense. I've been debating about using it in purple3...

**Review by grimmy — APPROVED** [review]
lgmt

**Comment by pabs3** [comment]
Please rewrite discord_rsa.c to use LF for line endings rather than CRLF.

**Comment by pabs3** [comment]
Please use https for the GNU URL in the discord_rsa.c license header.

**Comment by pabs3** [comment]
LLVM scan-build results are clear. My computer is too slow to run cppcheck though, if you are able to, please do that.

**Comment by EionRobb** [comment]
@pabs3 thanks for the suggestion.  I ran cppcheck over it, and no complaints from discord_rsa.c nor the changes in libdiscord.c (just some existing style complaints)

**Pull request merged**